### PR TITLE
Replace Eigen vectors of strings with std::vector<std::string>

### LINKED ISSUE #2864: stan_csv_reader.hpp uses Eigen vectors of strings
#### Summary:
The use of Eigen vectors of strings is unconventional and should be replaced by `std::vector<std::string>`.

#### Description:
Along the lines of what was done in #2863, `src/stan/io/stan_csv_reader.hpp` should replace uses of `Eigen::Matrix<std::string, Eigen::Dynamic, 1>` with `std::vector<std::string>`.

This affects only the `stan_csv::header` member and consequently the `stan_csv_reader::read_header()` method.

Upon cleanup, also one of the constructors in `src/stan/mcmc/chains.hpp` can be removed.

#### Additional Information:
The `stan::io::stan_csv_reader` class is referenced in CmdStan, so it should be reviewed if changes here will affect that.

#### Current Version:
v2.21.0

### PR DESCRIPTION
**Pull request by mcol**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary
Fixes #2864 as described in the subject and the issue. As a bonus already mentioned in the issue, the  `stan::mcmc::chains(const Eigen::Matrix<std::string, Dynamic, 1>&` constructor can be removed.

#### How to Verify
Existing tests should pass.

#### Side Effects
The type of `stan::io::stan_csv::header` has changed, but this is not referenced directly by any of our interfaces.

#### Documentation
None.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Marco Colombo

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/test/unit/io/stan_csv_reader_test.cpp [hunk 1]
```diff
@@ -106,66 +106,66 @@ TEST_F(StanIoStanCsvReader, read_metadata3) {
...
+  EXPECT_EQ("lp__", header[0]);
+  EXPECT_EQ("accept_stat__", header[1]);
+  EXPECT_EQ("stepsize__", header[2]);
+  EXPECT_EQ("treedepth__", header[3]);
+  EXPECT_EQ("n_leapfrog__", header[4]);
+  EXPECT_EQ("divergent__", header[5]);
+  EXPECT_EQ("energy__", header[6]);
+  EXPECT_EQ("d", header[7]);
+  EXPECT_EQ("sigmasq_delta", header[8]);
+  EXPECT_EQ("mu[1]", header[9]);
```
Comment (bob-carpenter) [inline src/test/unit/io/stan_csv_reader_test.cpp]: [optional]
These could go in a loop:

```
for (int i = 1; i <= 22; ++i) {
  std::stringstream ss;
  ss << "mu[" << i << "]";
  EXPECT_EQ(ss.str(), header[8 + i]);
}
```

Comment (mcol) [reply src/test/unit/io/stan_csv_reader_test.cpp]: We might just fix them here.

**Comment by mcol** [comment]
Thanks for the review! I'll merge as soon as tests pass.

**Comment by mcol** [comment]
This has had 3 consecutive random failures in Travis due to some error in the `apt-get` phase, so totally unrelated to the changes in this PR. How can we get past this?

**Comment by bob-carpenter** [comment]
@serban-nicusor-toptal and @seantalts --- what's the right path forward for this?  See the previous comment about apt-get failures.

**Comment by mcol** [comment]
The underlying apt-get issues should have been solved in #2868. If you know how to restart tests without pushing a new commit, perhaps it will go through.

**Comment by serban-nicusor-toptal** [comment]
@mcol Travis job restarted, hope it will take the new configuration.

**Comment by mcol** [comment]
Thanks @serban-nicusor-toptal, it has now passed the tests. Github tells me I'm not authorized to merge this PR, so someone else will have to do it for me.

**Comment by bob-carpenter** [comment]
I'm not sure why you wouldn't be approved to merge.  You're on the `maintainers` list, which is what is giving us merge permission.

**Comment by mcol** [comment]
I can merge on stan-dev/math, but here the button was shaded and the interface mentioned not being authorized... it's no big deal for me, though, as long as there are kind souls out there to do the merges for me. :)

**Comment by bob-carpenter** [comment]
Looks like Stan may be using separate permissions, so I added you to the stan team, too.  Let me know if you still have problems merging going forward.

**Pull request merged**